csrc/multidevice/communication.cpp

```diff
@@ -111,209 +98,225 @@ Broadcast::Broadcast(CommParams params) : Communication(params, "broadcast") {}
+          !params_.is_root_in_mesh) {
+        output_tensors.front().push_back(input_tensor);
+        continue;
+      }
+      output_tensors.front().push_back(output_tensor.slice(0, j, j + 1));
+      j++;
+    }
   }
-  return work;
+
+  return comm.getBackendForTeam(params_.team, backend)
+      ->gather(
+          output_tensors, input_tensors, {.rootRank = root_relative_index_});
 }
 
 Allgather::Allgather(CommParams params)
-    : Communication(params, "allgather", false) {
-  assertBufferCount(params_.src_bufs, 1);
-  assertBufferCount(params_.dst_bufs, params_.team.size());
-}
+    : Communication(params, "allgather", false) {}
 
 c10::intrusive_ptr<c10d::Work> Allgather::post(
     Communicator& comm,
+    at::Tensor input_tensor,
+    at::Tensor output_tensor,
     std::optional<CommunicatorBackend> backend) {
   post_common(*this, comm);
-  // This is used to change the representation of the buffers to match c10d
-  // ProcessGroup API
-  std::vector<std::vector<at::Tensor>> buf_list;
-  buf_list = {std::move(params_.dst_bufs)};
-  auto work = comm.getBackendForTeam(params_.team, backend)
-                  ->allgather(buf_list, params_.src_bufs, {});
-  params_.dst_bufs = std::move(buf_list.back());
-  return work;
-}
 
-Scatter::Scatter(CommParams params) : Communication(params, "scatter") {
-  assertBufferCount(params_.dst_bufs, 1);
+  std::vector<at::Tensor> input_tensors({input_tensor});
+
+  // FIXME: use split
+  std::vector<std::vector<at::Tensor>> output_tensors(1);
+  for (auto i : c10::irange(params_.team.size())) {
+    output_tensors[0].push_back(output_tensor.slice(0, i, i + 1));
+  }
+  return comm.getBackendForTeam(params_.team, backend)
+      ->allgather(output_tensors, input_tensors, {});
 }
 
+Scatter::Scatter(CommParams params) : Communication(params, "scatter") {}
+
 c10::intrusive_ptr<c10d::Work> Scatter::post(
     Communicator& comm,
+    at::Tensor input_tensor,
+    at::Tensor output_tensor,
     std::optional<CommunicatorBackend> backend) {
   post_common(*this, comm);
-  // This is used to change the representation of the buffers to match c10d
-  // ProcessGroup API
-  std::vector<std::vector<at::Tensor>> buf_list;
-  if (comm.deviceId() == params_.root) {
-    assertBufferCount(params_.src_bufs, params_.team.size());
-    buf_list = {std::move(params_.src_bufs)};
-  } else {
-    assertBufferCount(params_.src_bufs, 0);
+
+  if (comm.deviceId() == params_.root && !params_.is_root_in_mesh) {
+    output_tensor = at::empty_like(input_tensor.slice(0, 0, 1));
   }
-  auto work =
-      comm.getBackendForTeam(params_.team, backend)
-          ->scatter(
-              params_.dst_bufs, buf_list, {.rootRank = root_relative_index_});
+  std::vector<at::Tensor> output_tensors({output_tensor});
+
+  std::vector<std::vector<at::Tensor>> input_tensors;
   if (comm.deviceId() == params_.root) {
-    params_.src_bufs = std::move(buf_list.back());
+    input_tensors.resize(1);
```

Comment:
Why is input_tensor sized to 1 instead of `params_.team.size()`?

Reply:
`input_tensors` passed to `Backend::scatter` is a **vector of vectors** of tensors. In our current code, the first dimension is always 1. 

That being said, do you happen to know the reason behind that ProcessGroup API? Is that to allow multiple gathers to run in parallel? But I'd expect that to have to take multiple streams...

Reply:
Right....that argument decision confuses me everytime I see it...
I have no idea why they chose that? Maybe it's to merge multiple small scatter calls into a big one?

Reply:
https://github.com/pytorch/pytorch/blob/abdd569e412c4af5957b10ff8039511f8175986e/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L3381 is interesting... 

🤷